### ISSUE TITLE
TypeError: object HttpResponse can't be used in 'await' expression 

### ISSUE THREAD

**Issue by kojibhy** [django/channels#1558]
have notice issue when working with channels3 according to documentation for channels3:
```text
Then, adjust your project’s asgi.py file, e.g. myproject/asgi.py, to wrap the Django ASGI application:
import os
from channels.routing import ProtocolTypeRouter
from django.core.asgi import get_asgi_application
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "core.settings")
application = ProtocolTypeRouter({
    "http": get_asgi_application(),
    # Just HTTP for now. (We can add other protocols later.)
})
```
this works well in development but when you try switch to daphne server it start raise this error:

```python
2020-11-04 07:05:08,071 INFO     Starting server at tcp:port=8000:interface=0.0.0.0
2020-11-04 07:05:08,072 INFO     HTTP/2 support enabled
2020-11-04 07:05:08,072 INFO     Configuring endpoint tcp:port=8000:interface=0.0.0.0
2020-11-04 07:05:08,073 INFO     Listening on TCP address 0.0.0.0:8000
ERROR:django.request: Internal Server Error: /admin/login/ (2020-11-04 07:05:11; log.py:224)
Traceback (most recent call last):
  File ".../lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File ".../lib/python3.8/site-packages/django/utils/deprecation.py", line 116, in __call__
    response = self.process_response(request, response)
  File ".../lib/python3.8/site-packages/django/middleware/common.py", line 106, in process_response
    if response.status_code == 404:
AttributeError: 'coroutine' object has no attribute 'status_code'
2020-11-04 07:05:12,076 ERROR    Exception inside application: object HttpResponse can't be used in 'await' expression
Traceback (most recent call last):
  File ".../lib/python3.8/site-packages/django/core/handlers/asgi.py", line 161, in __call__
    response = await self.get_response_async(request)
  File ".../lib/python3.8/site-packages/django/core/handlers/base.py", line 148, in get_response_async
    response = await self._middleware_chain(request)
TypeError: object HttpResponse can't be used in 'await' expression
127.0.0.1:41488 - - [04/Nov/2020:07:05:12] "GET /admin/login/?next=/admin/oloo/ownermodel/" 500 453
```
to fix this need to write something like this, but as I understand this is a fallback for 2.2 Django but in my env, I have the latest Django.

```python
import os
import django
from channels.http import AsgiHandler
from channels.routing import ProtocolTypeRouter, URLRouter
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "core.settings")
django.setup()
from apps.ws_routing import websocket_urlpatterns

application = ProtocolTypeRouter({
    "http": AsgiHandler(),
    'websocket':
        URLRouter(
            websocket_urlpatterns
        )
})

```

Django==3.1.3
asgiref==3.3.0
daphne==3.0.0
channels==3.0.0

**Comment by carltongibson** [comment]
Hi @kojibhy — This looks like an error on the middleware stack. The issue is coming up inside Django's `ASGIHandler`...

```
  File ".../lib/python3.8/site-packages/django/core/handlers/base.py", line 148, in get_response_async
    response = await self._middleware_chain(request)
TypeError: object HttpResponse can't be used in 'await' expression
```

I think this would come up if you took channels out of the equation entirely: 

```
import os
from django.core.asgi import get_asgi_application
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "core.settings")
application = get_asgi_application()
```

Which is ≈just Django's default `asgi.py`. 

Not enough here to see exactly what's going on. What are your middleware? 

Can you provide a sample project that reproduces this?

**Comment by kojibhy** [comment]
Hi, event after changing to asgi you provide it still raise these issues.
my middleware classes will share url soon with project
```python
MIDDLEWARE = [
    'corsheaders.middleware.CorsMiddleware',
    'django_hosts.middleware.HostsRequestMiddleware',
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django_hosts.middleware.HostsResponseMiddleware',
]

```

**Comment by carltongibson** [comment]
@kojibhy Can you remove this: 

```
    'corsheaders.middleware.CorsMiddleware',
    'django_hosts.middleware.HostsRequestMiddleware',
```

i.e. the non-Django middleware. 

It's probably one of them needing an update to be async-aware.

**Comment by kojibhy** [comment]
it`s true they are both raise issues.

but it seems they are both support Django 3

- CorsMiddleware(MiddlewareMixin)
- HostsBaseMiddleware(MiddlewareMixin)

**Comment by carltongibson** [comment]
Super. No problem. Everyone is still updating together so happy to look at it. I'll close this now. 👍

**Comment by marydiane** [comment]
How didyou fix this object HttpResponse can't be used in 'await' expression

**Comment by jamietdavidson** [comment]
If anyone stumbles on this thread and they're trying to integrate with strawberry (this might apply to other configurations, too). I found that removing `"daphne"` from `INSTALLED_APPS` made everything work as expected.

**Comment by carltongibson** [comment]
> If anyone stumbles on this thread and they're trying to integrate with strawberry (this might apply to other configurations, too). I found that removing "daphne" from INSTALLED_APPS made everything work as expected.

This would mean reverting to WSGI, rather than addressing the issue.

**Comment by jamietdavidson** [comment]
@carltongibson the channels docs mention that daphne is optional - am I missing something? Also, if this is the case, I cannot for the life of me get this to work. Is this a problem with the core library that needs to be address? 

`response = await self._middleware_chain(request)` is the exact stack trace I am receiving, too

**Comment by carltongibson** [comment]
@jamietdavidson how about opening a fresh issue with a full description and reproduce?

**Comment by onyekaa** [comment]
Just ran into this error, even for inbuilt middleware, and I resolved it by upgrading to Django > 4.1.2. Seems the sync/async fallback Django has for this wasn't added till then? And also our custom middlewares inherit the MiddleWareMixin, if any of that helps @jamietdavidson
